ApiData.ts

```diff
@@ -0,0 +1,264 @@
+		}
+	}
+
+	/** Update the page information and records display based on the current state of the grid. */
+	private updatePageInfo(): void {
+		const totalPages = Math.ceil(this.totalItems / this.pageSize);
+		const pageInfo = `Page ${this.currentPage} of ${totalPages}`;
+		const from = this.firstVal;
+		let to = Math.min(from + this.pageSize, this.maxRange);
+		$('#pageInfo').text(`${pageInfo}`);
+		$('.records').text(`Showing records ${from} to ${to}`);
+	}
+
+	private setupControls(): void {
+		$('#prevBtn').on('click', () => this.handlePageChange(-1));
+		$('#nextBtn').on('click', () => this.handlePageChange(1));
+		$(window).on('resize', debounce(() => { this.handleResize(); }, 100));
+	}
+
+	/** Handles page navigation by updating the firstVal, lastVal, current page, and enabling/disabling previous and next buttons as needed. */
+	private handlePageChange(delta: number): void {
+		let prevBtn = $('#prevBtn');
+		let nextBtn = $('#nextBtn');
+
+		// Check if delta(change in page number) is positive and disable the next page if firstval + pageSize exceeds the MaxRange.
+		if (delta > 0 && this.firstVal + delta * this.pageSize > this.maxRange) {
+			this.firstVal = this.lastVal - this.pageSize;
+			prevBtn.attr("disabled", null);
+			nextBtn.attr("disabled", "disabled");
+		} else if (delta < 0 && this.firstVal + delta * this.pageSize < 0) {
+			// If delta is negative then reset firstVal to 0 and disable prev button 
+			this.firstVal = 0;
+			prevBtn.attr("disabled", "disabled");
+			nextBtn.attr("disabled", null);
+		} else {
+			this.firstVal = Math.max(0, Math.min(this.firstVal + delta * this.pageSize, this.maxRange));
+			prevBtn.attr("disabled", null);
+			nextBtn.attr("disabled", null);
+		}
+
+		this.lastVal = this.firstVal + delta * this.pageSize;
+		this.currentPage = Math.floor(this.firstVal / this.pageSize) + 1;
+
+		this.fetchAndDisplayRecords()
+			.then(() => {
+				this.updatePageInfo();
+			})
+			.catch(error => {
+				console.error("Error fetching records while changing page :", error);
+				alert('Error occured while changing page!');
+			});
+	}
+
+	private handleResize(): void {
+		const newGridSize = Math.floor((Math.floor(<number>($(window).innerHeight())) * GRID_RATIO) / ROW_HEIGHT) - 1;
+
+		// Check if the new grid size is non-negative
+		if (newGridSize >= 0) {
+			// Adjust firstVal for the last page
+			if (this.firstVal + newGridSize > this.maxRange) {
+				this.firstVal = this.maxRange - newGridSize;
+			}
+
+			this.pageSize = newGridSize;
+			this.lastVal = this.firstVal + newGridSize;
+
+			this.adjustGridHeight();
+
+			this.fetchAndDisplayRecords()
+				.then(() => {
+					this.updatePageInfo();
+				})
+				.catch(error => {
+					console.error("Error fetching records while resizing:", error);
+					alert('Error occured while resizing!');
+				});
+		}
+	}
+
+	private displayRecords(): void {
+		const gridTemplate = new GridTemplate(this.columnNames, this.data);
```

Comment:
So we are creating the GridTemplate class everytime we displayRecords()?

I would have loved that we could just set the dataRecords every time we update the display because `columnNames` never changes throughout this life. It just sounds like a lot of objects being created and created and created...